docs/specs/browserResidentDesktopAgents.md

```diff
@@ -0,0 +1,224 @@
+
+> Note - This covers most typical cases but it leaves out more complex launching scenarios such as "daisy chains" (frame -> frame or open -> open) or allowing end users to open tabs in an adhoc manner. To cover these cases, DAs may choose to allow themselves to be opened as hidden iframes within app windows. See the section on hidden iframes at the end of this document.
+
+## Responding to app instance connections - Web Connection Protocol (WCP)
+
+DAs MUST run `addEventListener("message",...)` to receive incoming connection requests from apps. This will receive the initial `postMessage` "Handshake" messages from apps. Handshakes do not contain any identifying data. They contain only a "nonce" (a unique id).
+
+Upon receiving an incoming Handshake the DA MUST:
+
+1) Create a [MessageChannel](https://developer.mozilla.org/en-US/docs/Web/API/Channel_Messaging_API)
+2) Set up a processing thread to receive and process BCP requests for this connection
+3) Respond to the event's `source` window by echoing the nonce and transferring `port2` from the channel
+
+Example WCP negotiation
+```JavaScript
+window.addEventListener("message", (event) => {
+    const { origin, data, source } = event
+    const channel = new MessageChannel()
+    startBCPProcessing(event, channel) // See next step
+    source.postMessage({ nonce: data.nonce }, origin, [channel.port2])
+});
+```
+
+Alternatively for step 3, a DA MAY return a `url` instead transferring a MessagePort. When `url` is provided in the response, `getAgent()` will open that `url` in a hidden iframe and initiate another WCP handshake with that frame.
+
+Example WCP negotiation with URL
+```JavaScript
+window.addEventListener("message", (event) => {
+    const { origin, data, source } = event
+    source.postMessage({ nonce: data.nonce, url }, origin)
+});
+```
+
+## Responding to app communications with Browser Communication Protocol (BCP)
+
+BCP processing should begin by setting up an inactive connection instance. This instance will become active after the first BCP "WCPValidateAppIdentity" message is received and processed (or deleted if it fails). It is important to remember the WindowProxy (event.source).
+
+An `instanceUuid` should be established at this point. This will be used in following steps.
+
+```JavaScript
+const startBCPProcessing(event, channel) => {
+    const instanceUuid = uuid();
+    const connection = {
+        status: "inactive",
+        windowProxy: event.source,
+        origin: event.origin,
+        channel,
+        instanceUuid
+    };
+    connections[instanceUuid] = connection;
+    channel.port1.onmessage = (e) => {
+        processIncomingBCP(connection, e); // This function should switch based on message type
+    }
+}
+```
+
+### Step 1 - Validating a connection
+
+The first BCP message received should be a "WCPValidateAppIdentity" message. With this message, the DA establishes or denies the connection.
```

Comment:
This is where we differ a little from the DAB protocol, this is post-handshake so that it happens on the final comms channel (MessagePort) rather than Window.postMessage (which is not a unique channel).